internal/collector/otel_collector_plugin.go

```diff
@@ -117,19 +121,190 @@ func (oc *Collector) Close(ctx context.Context) error {
+	nginxConfigContext, ok := msg.Data.(*model.NginxConfigContext)
+	if !ok {
+		slog.ErrorContext(ctx, "Unable to cast message payload to *model.NginxConfigContext", "payload", msg.Data)
+	}
+
+	reloadCollector := oc.checkForNewNginxReceivers(nginxConfigContext)
+
+	if reloadCollector {
+		slog.InfoContext(ctx, "Reloading OTel collector config")
+		err := writeCollectorConfig(oc.config.Collector)
+		if err != nil {
+			slog.ErrorContext(ctx, "Failed to write OTel Collector config", "error", err)
+			return
+		}
+
+		oc.restartCollector(ctx)
+	}
+}
+
+func (oc *Collector) restartCollector(ctx context.Context) {
+	err := oc.Close(ctx)
+	if err != nil {
+		slog.ErrorContext(ctx, "Failed to shutdown OTel Collector", "error", err)
+		return
+	}
+
+	settings := OTelCollectorSettings(oc.config)
+	oTelCollector, err := otelcol.NewCollector(settings)
+	if err != nil {
+		slog.ErrorContext(ctx, "Failed to create OTel Collector", "error", err)
+		return
+	}
+	oc.service = oTelCollector
+
+	var runCtx context.Context
+	runCtx, oc.cancel = context.WithCancel(ctx)
+
+	go func() {
+		bootErr := oc.bootup(runCtx)
+		if bootErr != nil {
+			slog.ErrorContext(runCtx, "Unable to start OTel Collector", "error", bootErr)
+		}
+	}()
+}
+
+func (oc *Collector) checkForNewNginxReceivers(nginxConfigContext *model.NginxConfigContext) bool {
+	nginxReceiverFound, reloadCollector := oc.updateExistingNginxPlusReceiver(nginxConfigContext)
+
+	if !nginxReceiverFound && nginxConfigContext.PlusAPI != "" {
+		oc.config.Collector.Receivers.NginxPlusReceivers = append(
+			oc.config.Collector.Receivers.NginxPlusReceivers,
+			config.NginxPlusReceiver{
+				InstanceID: nginxConfigContext.InstanceID,
+				PlusAPI:    nginxConfigContext.PlusAPI,
+			},
+		)
+
+		reloadCollector = true
+	} else if nginxConfigContext.PlusAPI == "" {
+		nginxReceiverFound, reloadCollector = oc.updateExistingNginxOSSReceiver(nginxConfigContext)
+
+		if !nginxReceiverFound && nginxConfigContext.StubStatus != "" {
+			oc.config.Collector.Receivers.NginxReceivers = append(
+				oc.config.Collector.Receivers.NginxReceivers,
+				config.NginxReceiver{
+					InstanceID: nginxConfigContext.InstanceID,
+					StubStatus: nginxConfigContext.StubStatus,
+				},
+			)
+
+			reloadCollector = true
+		}
+	}
+
+	return reloadCollector
+}
+
+func (oc *Collector) updateExistingNginxPlusReceiver(
+	nginxConfigContext *model.NginxConfigContext,
+) (nginxReceiverFound, reloadCollector bool) {
+	reloadCollector = false
+	nginxReceiverFound = false
+
+	for index, nginxPlusReceiver := range oc.config.Collector.Receivers.NginxPlusReceivers {
+		if nginxPlusReceiver.InstanceID == nginxConfigContext.InstanceID {
+			nginxReceiverFound = true
+
+			if nginxPlusReceiver.PlusAPI != nginxConfigContext.PlusAPI {
+				oc.config.Collector.Receivers.NginxPlusReceivers = append(
+					oc.config.Collector.Receivers.NginxPlusReceivers[:index],
+					oc.config.Collector.Receivers.NginxPlusReceivers[index+1:]...,
```

Comment:
Won't this cause a "slice bounds out of range" panic if `nginxPlusReceiver` is the last element in `oc.config.Collector.Receivers.NginxPlusReceivers`?

Reply:
I dont think so I have a unit test here that passes where I remove the receiver from the slice https://github.com/nginx/agent/blob/intro-nginx-plus-receiver/internal/collector/otel_collector_plugin_test.go#L288

Reply:
I tested this out in Go playground and it does **not** go out of bounds 👍 

So if you subslice with `slice[len(slice):]`, the result is an empty slice, but if you do `slice[len(slice)+1:]` you get the "slice out of bounds" error. Learn something new every day.